File: test-suite/src/test/java/io/micronaut/gcp/pubsub/subscriber/ReactivePushSubscriberTest.java
package io.micronaut.gcp.pubsub.subscriber;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.dataformat.xml.XmlMapper;
import com.google.pubsub.v1.PubsubMessage;
import io.micronaut.context.annotation.Property;
import io.micronaut.gcp.pubsub.push.PushRequest;
import io.micronaut.gcp.pubsub.support.Animal;
import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.HttpStatus;
import io.micronaut.http.client.HttpClient;
import io.micronaut.http.client.annotation.Client;
import io.micronaut.json.JsonMapper;
import io.micronaut.test.annotation.MockBean;
import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Inject;
import jakarta.inject.Named;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import reactor.core.publisher.Mono;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.HashMap;

@MicronautTest
@Property(name = "spec.name", value = "ReactivePushSubscriberTest")
@Property(name = "gcp.projectId", value = "test-project")
class ReactivePushSubscriberTest {

    @Inject
    @Client("/")
    HttpClient pushClient;

    @Inject
    @Named("xml")
    XmlMapper xmlMapper;

    @Inject
    JsonMapper jsonMapper;

    Object unwrappedResult;

    @BeforeEach
    void setup() {
        unwrappedResult = null;
    }

    @Test
    void testRawBytes() {
        byte[] bytesSent = "foo".getBytes(StandardCharsets.UTF_8);
        String encodedData = Base64.getEncoder().encodeToString(bytesSent);
        PushRequest request = new PushRequest("projects/test-project/subscriptions/raw-push-subscription", new PushRequest.PushMessage(new HashMap<>(), encodedData, "1", "2021-02-26T19:13:55.749Z"));
        HttpResponse<?> response = pushClient.toBlocking().exchange(HttpRequest.POST("/push", request));

        Assertions.assertEquals(HttpStatus.OK, response.getStatus());
        Assertions.assertNotNull(unwrappedResult);
        Assertions.assertInstanceOf(byte[].class, unwrappedResult);
        byte[] convertedResult = (byte[]) unwrappedResult;
        String decodedMessage = new String(convertedResult, StandardCharsets.UTF_8);
        Assertions.assertEquals("foo", decodedMessage);
    }

    @Test
    void testNativeMessage() {
        byte[] bytesSent = "foo".getBytes(StandardCharsets.UTF_8);
        String encodedData = Base64.getEncoder().encodeToString(bytesSent);
        PushRequest request = new PushRequest("projects/test-project/subscriptions/native-push-subscription", new PushRequest.PushMessage(new HashMap<>(), encodedData, "1", "2021-02-26T19:13:55.749Z"));
        HttpResponse<?> response = pushClient.toBlocking().exchange(HttpRequest.POST("/push", request));

        Assertions.assertEquals(HttpStatus.OK, response.getStatus());
        Assertions.assertNotNull(unwrappedResult);
        Assertions.assertInstanceOf(PubsubMessage.class, unwrappedResult);
        PubsubMessage convertedResult = (PubsubMessage) unwrappedResult;
        String decodedMessage = convertedResult.getData().toString(StandardCharsets.UTF_8);
        Assertions.assertEquals("foo", decodedMessage);
    }


    @Test
    void testJsonPojo() throws IOException {
        Animal dog = new Animal("dog");
        String encodedData = Base64.getEncoder().encodeToString(jsonMapper.writeValueAsBytes(dog));
        PushRequest request = new PushRequest("projects/test-project/subscriptions/animals-push", new PushRequest.PushMessage(new HashMap<>(), encodedData, "1", "2021-02-26T19:13:55.749Z"));
        HttpResponse<?> response = pushClient.toBlocking().exchange(HttpRequest.POST("/push", request));

        Assertions.assertEquals(HttpStatus.OK, response.getStatus());
        Assertions.assertNotNull(unwrappedResult);
        Assertions.assertInstanceOf(Animal.class, unwrappedResult);
        Animal convertedResult = (Animal) unwrappedResult;
        Assertions.assertEquals("dog", convertedResult.getName());
    }

    @Test
    void testXmlPojo() throws JsonProcessingException {
        Animal dog = new Animal("cat");
        String encodedData = Base64.getEncoder().encodeToString(xmlMapper.writeValueAsBytes(dog));
        PushRequest request = new PushRequest("projects/test-project/subscriptions/animals-legacy-push", new PushRequest.PushMessage(new HashMap<>(), encodedData, "1", "2021-02-26T19:13:55.749Z"));
        HttpResponse<?> response = pushClient.toBlocking().exchange(HttpRequest.POST("/push", request));

        Assertions.assertEquals(HttpStatus.OK, response.getStatus());
        Assertions.assertNotNull(unwrappedResult);
        Assertions.assertInstanceOf(Animal.class, unwrappedResult);
        Animal convertedResult = (Animal) unwrappedResult;
        Assertions.assertEquals("cat", convertedResult.getName());
    }

    @MockBean(MessageProcessor.class)
    MessageProcessor mockMessageProcessor() {
        return new MessageProcessor() {
            @Override
            public Mono<Boolean> handleByteArrayMessage(byte[] message) {
                unwrappedResult = message;
                return MessageProcessor.super.handleByteArrayMessage(message);
            }

            @Override
            public Mono<Boolean> handlePubSubMessage(PubsubMessage pubsubMessage) {
                unwrappedResult = pubsubMessage;
                return MessageProcessor.super.handlePubSubMessage(pubsubMessage);
            }

            @Override
            public Mono<Boolean> handleAnimalMessage(Animal message) {
                unwrappedResult = message;
                return MessageProcessor.super.handleAnimalMessage(message);
            }
        };
    }
}

File: test-suite-python/src/test/python/micronaut/gcp/pubsub/subscriber/ReactivePushSubscriberTest.py
from typing import Annotated

import java
from com.google.pubsub.v1 import PubsubMessage
from jakarta.inject import Inject, Named
from java.util import Base64, HashMap
from micronaut.context.annotation import Property
from micronaut.gcp.pubsub.push import PushRequest
from micronaut.gcp.pubsub.support.Animal import Animal
from micronaut.http import HttpRequest, HttpStatus
from micronaut.http.client import HttpClient
from micronaut.http.client.annotation import Client
from micronaut.json import JsonMapper
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import BeforeEach, Test
from tools.jackson.dataformat.xml import XmlMapper

from micronaut.gcp.pubsub.subscriber.MessageProcessor import MessageProcessor


@MicronautTest
@Property(name="spec.name", value="ReactivePushSubscriberTest")
@Property(name="gcp.projectId", value="test-project")
class ReactivePushSubscriberTest:
    pushClient: Annotated[HttpClient, Inject, Client("/")]
    xmlMapper: Annotated[XmlMapper, Inject, Named("xml")]
    jsonMapper: Annotated[JsonMapper, Inject]
    messageProcessor: Annotated[MessageProcessor, Inject]

    @BeforeEach
    def setup(self):
        self.messageProcessor.unwrapped_result = None

    def push(self, subscription: str, encoded_data: str):
        request = PushRequest("projects/test-project/subscriptions/" + subscription,
                              PushRequest.PushMessage(HashMap(), encoded_data, "1", "2021-02-26T19:13:55.749Z"))
        return self.pushClient.toBlocking().exchange(HttpRequest.POST("/push", request))

    @Test
    def test_raw_bytes(self):
        response = self.push("raw-push-subscription", Base64.getEncoder().encodeToString("foo".encode()))

        assert response.getStatus() == HttpStatus.OK
        result = self.messageProcessor.unwrapped_result
        assert result is not None
        assert bytes(result).decode() == "foo"

    @Test
    def test_native_message(self):
        response = self.push("native-push-subscription", Base64.getEncoder().encodeToString("foo".encode()))

        assert response.getStatus() == HttpStatus.OK
        result = self.messageProcessor.unwrapped_result
        assert java.instanceof(result, PubsubMessage)
        assert result.getData().toStringUtf8() == "foo"

    @Test
    def test_json_pojo(self):
        response = self.push("animals-push", Base64.getEncoder().encodeToString(self.jsonMapper.writeValueAsBytes(Animal("dog"))))

        assert response.getStatus() == HttpStatus.OK
        result = self.messageProcessor.unwrapped_result
        assert result is not None
        assert result.name == "dog"

    @Test
    def test_xml_pojo(self):
        response = self.push("animals-legacy-push", Base64.getEncoder().encodeToString(self.xmlMapper.writeValueAsBytes(Animal("cat"))))

        assert response.getStatus() == HttpStatus.OK
        result = self.messageProcessor.unwrapped_result
        assert result is not None
        assert result.name == "cat"
